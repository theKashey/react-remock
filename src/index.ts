import {createElement, cloneElement, enable, disable, patchJsxRuntime} from "./patch/react";

export * from "./exports";
export {
  createElement,
  cloneElement,
  enable,
  disable,
}
