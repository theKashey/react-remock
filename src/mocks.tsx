import * as React from 'react';

export type AnyElement = string | React.ComponentType<any>;
export type ElementMatcher = AnyElement | RegExp;
export type PartialMockedElement = { type?: AnyElement, props?: any, children?: any[] };

export type Matcher = (type: ElementMatcher, props: any, children?: any[]) => boolean;
export type Mocker = (type: AnyElement, props?: any, children?: any[]) => undefined | PartialMockedElement | React.ReactElement<any>
export type Unsubscribe = () => void;

export interface Mock {
  test: Matcher;
  replace: Mocker;
}

let matchers: Mock[] = [];

const getNameOf = (type: any): string => {
  if (!type) return 'Unknown';
  if (typeof type === 'string') return type;
  return type.displayName || type.name || 'Component';
};

const realAddMock = (mock: Mock): Unsubscribe => {
  matchers.push(mock);
  return () => matchers = matchers.filter(el => el !== mock);
};

const MockedOut: React.SFC = () => null;
const MockedTransparent: React.SFC = ({children}) => children as any;

const defaultMock = (type: any, props: any) => ({
  type: MockedOut,
  props: {
    itWas: getNameOf(type),
    ...props
  }
});

const mock = (type: ElementMatcher, mockBy?: Mocker): Unsubscribe => {
  const mock = {
    test: (element: AnyElement) => {
      const elementName = getNameOf(element);
      return element === type ||
        (
          typeof type !== 'function' && (
            elementName === type ||
            (typeof type !== "string" && !!elementName.match(type))
          )
        )
    },
    replace: mockBy || defaultMock
  };
  return realAddMock(mock);
};

const match = (match: Matcher, mockBy?: Mocker): Unsubscribe => {
  const mock = {
    test: match,
    replace: mockBy || defaultMock
  };
  return realAddMock(mock);
};

const transparent = (type: ElementMatcher): Unsubscribe => {
  return mock(type, () => ({type: MockedTransparent}))
};

const renderProp = (type: ElementMatcher, ...args: any[]): Unsubscribe => {
  return mock(
    type,
    (type, props, _children) => {
      const children = Array.isArray(_children) ? _children[0] : _children;

      if (typeof children !== "function") {
        console.error('remock: mocked', type, ' expects', children, ' to be a function (it is ', typeof children, ')');
        return {};
      }
      return {
        type: MockedTransparent,
        props: {
          itWas: getNameOf(type),
          ...props,
        },
        children: (children as any)(...args)
      }
    }
  )
};

const unmock =
  (type: AnyElement, props?: any) => {
    matchers = matchers.filter(mock => !mock.test(type, props))
  }

const clearMocks = () => {
  matchers = [];
};

const getMocks = () => matchers;


export {
  mock,
  match,
  unmock,

  transparent,
  renderProp,

  clearMocks,

  getMocks
}