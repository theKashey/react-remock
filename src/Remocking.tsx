import * as React from 'react';
import {ElementMatcher, Mocker, mock} from "./mocks";

export interface Props {
  component: ElementMatcher;
  by?: Mocker;
}

export class Remocking extends React.Component<Props> {
  // remock is adding the mock in `resolver`

  componentWillUnmount() {
    (this.props as any).__unmock();
  }

  render(): React.ReactNode {
    return this.props.children
      ? React.Children.only(this.props.children)
      : null;
  }
}