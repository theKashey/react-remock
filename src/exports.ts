import * as remock from './mocks';
import {Remocking} from './Remocking';

export {
  remock,
  Remocking,
}