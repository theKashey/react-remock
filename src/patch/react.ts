import ReactDefault from 'react';
import * as ReactAll from 'react';
import {resolver} from '../resolver';

const SIGN = 'isPatchedByReactRewiremock';

// to support old jest tests use two different types of import
const React = ReactDefault || ReactAll;

export const createElement = React.createElement;
export const cloneElement = React.cloneElement;
export const createFactory = React.createFactory;

/**
 * Wraps jsx factory with a mock
 */
const wrapJsx = <T extends typeof React.createElement>(jsxFactory: T): T => (
    (type: any, props: any = {}, ...args: any[]) => {
        const anyProps = props || {};
        const resolved = resolver(type, props, args) as any;
        // it could be resolved to a non-object, and that's legit for renderProp case
        if (resolved === undefined) {
            // @ts-ignore
            return jsxFactory.apply(this, arguments);
        }
        if (!resolved) {
            return resolved;
        }
        const {type: newType = type, props: newProps = props, children = args} = resolved;
        const key = anyProps.key;
        const ref = anyProps.ref;
        const finalProps = {
            key,
            ref,
            ...newProps
        };

        return finalProps && (finalProps.children !== anyProps.children)
            ? jsxFactory(newType, finalProps)
            : jsxFactory(newType, finalProps, ...children);
    }) as any as T;

/**
 * patches jsx-runtime
 * @example
 * ```tsx
 * import jsxRuntime from "react/jsx-runtime";
 * import jsxRuntimeDev from "react/jsx-dev-runtime";
 *
 * import {patchJsxRuntime} from 'react-remock';
 *
 * patchJsxRuntime(jsxRuntime);
 * patchJsxRuntime(jsxRuntimeDev);
 * ```
 */
export const patchJsxRuntime = <T extends {jsx:any; jsxs?:any; jsxDEV?:any}>(runtime:T): T => {
    // inspiration from https://github.com/preactjs/signals/blob/main/packages/react/runtime/src/auto.ts#L338
    runtime.jsx && /*   */ (runtime.jsx = wrapJsx(runtime.jsx));
    runtime.jsxs && /*  */ (runtime.jsxs = wrapJsx(runtime.jsxs));
    runtime.jsxDEV && /**/ (runtime.jsxDEV = wrapJsx(runtime.jsxDEV));

    return runtime;
}

function patchReact(React: any) {
    if (React) {
        if (!React.createElement[SIGN]) {

            React.createElement = wrapJsx(createElement);

            React.cloneElement =
                (type: any, props: any, ...args: any[]) => {
                    const newElement = cloneElement(type, props, ...args);
                    const key = newElement.key;
                    const ref = newElement.ref;
                    // extract children to separate prop
                    const {children = [], ...rest} = newElement.props || {};
                    return React.createElement(newElement.type, {
                        key,
                        ref,
                        ...rest
                    }, ...children);
                };

            React.createFactory = (type: any) => {
                const factory = React.createElement.bind(null, type);
                factory.type = type;
                return factory;
            }

            Object.defineProperty(React.createElement, SIGN, {
                configurable: false,
                writable: false,
                enumerable: false,
                value: true,
            });
        } else {
            console.log('skipping double patching');
        }
    }
}

export const enable = () => patchReact(React);
export const disable = () => {
    const R = React as any;
    R.createElement = createElement;
    R.cloneElement = cloneElement;
    R.createFactory = createFactory;
};

enable();