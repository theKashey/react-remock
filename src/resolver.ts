import * as React from 'react';
import {AnyElement, getMocks, mock, PartialMockedElement} from "./mocks";
 import {Remocking} from './Remocking';

const resolver = (type: AnyElement, props: any, children: any[]): PartialMockedElement | React.ReactElement<any> => {
  if (type === Remocking) {
     props.__unmock = mock(props.component, props.by);
  }
  const mocks = getMocks().filter(match => match.test(type, props || {}, children));
  if (mocks.length) {
    return mocks[mocks.length - 1].replace(type, props, children)
  }
  return undefined;
};


export {
  resolver
}