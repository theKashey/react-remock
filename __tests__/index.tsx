import * as React from 'react';
import {shallow, mount} from 'enzyme';
import {remock, Remocking, enable, disable} from '../src';

const ComponentRed = () => <React.Fragment>Red</React.Fragment>;
const ComponentBlue = () => <React.Fragment>Blue</React.Fragment>;
const ComponentBlack = () => <React.Fragment>Black</React.Fragment>;
const Color: React.SFC<{ color: string }> = ({color}) => <React.Fragment>{color}</React.Fragment>;

describe('Remock', () => {

  describe('mock', () => {
    it('test before all', () => {
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
      </div>);
      expect(wrapper.html()).toBe("<div>RedBlueRedBlueRedBlue</div>");
    });

    it('Should mock all red ones by name', () => {
      remock.mock('ComponentRed');
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>BlueBlueBlue</div>");
    });

    it('Should unmock', () => {
      const un1=remock.mock('ComponentRed');
      const un2=remock.mock('ComponentBlue');
      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>Black</div>");
      un1();
      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>RedBlack</div>");
      un2();
      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>RedBlueBlack</div>");
      remock.clearMocks();
    });

    it('Should not mock string without exact match', () => {
      remock.mock('Component');
      remock.mock('componentred');
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>RedBlueBlack</div>");
    });

    it('Should mock all red ones by RegExp', () => {
      remock.mock(/Component(Red|Blue)/);
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>Black</div>");
    });


    it('Should mock all Blue ones by Class', () => {
      remock.mock(ComponentBlue);
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>RedRedRed</div>");
    });

    it('Should replace all Blue ones by Black', () => {
      remock.mock(ComponentBlue, () => ({type: Color, props: {color: 'Black'}}));
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>RedBlackRedBlackRedBlack</div>");
    });

    it('Should replace children', () => {
      remock.mock(ComponentBlue, (type, props) => <span {...props}>mocked</span>);
      const wrapper = shallow(<div>
        <ComponentBlue>original</ComponentBlue>
        <ComponentBlue>child</ComponentBlue>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div><span>mocked</span><span>mocked</span></div>");
    });

    it('enable/disable', () => {
      remock.mock(/Component(Red|Blue)/);

      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>Black</div>");
      
      disable();

      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>RedBlueBlack</div>")

      enable();


      expect(shallow(<div>
        <ComponentRed/><ComponentBlue/><ComponentBlack/>
      </div>).html()).toBe("<div>Black</div>")

      remock.clearMocks();
    });

    it('should transparentize', () => {
      const BadComponent: React.SFC = () => <span>dead end</span>;

      const wrapper1 = mount(<BadComponent>good content</BadComponent>);
      expect(wrapper1.html()).toBe("<span>dead end</span>");

      remock.transparent(BadComponent);
      const wrapper2 = mount(<BadComponent>
        <div>good content</div>
      </BadComponent>);
      remock.clearMocks();

      expect(wrapper2.html()).toBe("<div>good content</div>");
    });

    it('should unwrap renderprop', () => {
      const RenderProp: React.SFC<{ children: (a: number) => React.ReactElement<any> }> = ({children}) => children(42);

      const wrapper1 = mount(<RenderProp>{x => <div>{x}</div>}</RenderProp>);
      expect(wrapper1.html()).toBe("<div>42</div>");

      const wrapper2 = shallow(<RenderProp>{x => <div>{x}</div>}</RenderProp>);
      expect(wrapper2.html()).toBe("<div>42</div>");

      remock.renderProp(RenderProp, 24);
      const wrapper3 = mount(<RenderProp>{x => <div>{x}</div>}</RenderProp>);
      remock.clearMocks();
      expect(wrapper3.html()).toBe("<div>24</div>");
      expect(wrapper3.debug()).toMatchSnapshot()
    });

    it('should unwrap class renderprop', () => {

      class RenderProp extends React.Component<{ children: (a: number) => React.ReactElement<any> }> {
        render() {
          return <div>will render {this.props.children(42)}</div>;
        }
      }

      const TopLevel: React.SFC = () => <RenderProp>{x => <div>{x}</div>}</RenderProp>;

      const wrapper1 = mount(<TopLevel/>);
      expect(wrapper1.text()).toBe("will render 42");

      const wrapper2 = shallow(<TopLevel/>);
      expect(wrapper2.text()).toBe("<RenderProp />");

      remock.renderProp(RenderProp, 24);
      const wrapper3 = mount(<TopLevel/>);
      remock.clearMocks();
      expect(wrapper3.text()).toBe("24");
    });

    // ---------------------------

    it('test after all', () => {
      const wrapper = shallow(<div>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
        <ComponentRed/><ComponentBlue/>
      </div>);
      expect(wrapper.html()).toBe("<div>RedBlueRedBlueRedBlue</div>");
    });
  });

  describe('match', () => {
    it('Should match by prop', () => {
      remock.match((type, props = {}) => props.color === "red", () => ({props: {color: "match"}}));
      const wrapper = shallow(<div>
        <Color color="blue"/><Color color="red"/><Color color="green"/>
      </div>);
      remock.clearMocks();
      expect(wrapper.html()).toBe("<div>bluematchgreen</div>");
    });
  });

  describe('Remock', () => {
    it('Should mock red', () => {
      const wrapper1 = mount(
        <div>
          <Remocking component="ComponentRed"/>

          <ComponentRed/>
          <ComponentBlue/>

          <Remocking component={ComponentBlue}/>
        </div>
      );

      expect(wrapper1.html()).toBe("<div>Blue</div>");
      wrapper1.unmount();
    });

    it('Should mock blue', () => {
      const wrapper2 = mount(
        <div>
          <Remocking component={ComponentBlue}/>

          <ComponentRed/>
          <ComponentBlue/>
        </div>
      );
      expect(wrapper2.html()).toBe("<div>Red</div>");
      wrapper2.unmount();
    });
  });

  describe('React API', () => {
    describe('createElement', () => {
      it('replace type', () => {
        remock.mock('ComponentRed', () => ({type: ComponentBlue}));
        const test = React.createElement(ComponentRed, {prop1: 1}, 42);
        remock.clearMocks();
        expect(test.type).toBe(ComponentBlue);
        expect(test.props).toEqual({
          prop1: 1,
          children: 42
        });
      });

      it('replace prop', () => {
        remock.mock('ComponentRed', () => ({props: {prop2: 2}}));
        const test = React.createElement(ComponentRed, {prop1: 1}, 42);
        remock.clearMocks();
        expect(test.props).toEqual({
          prop2: 2,
          children: 42
        })
      });

      it('merge prop', () => {
        remock.mock('ComponentRed', (type, props) => ({props: {...props, prop2: 2}}));
        const test = React.createElement(ComponentRed, {prop1: 1, key: 'key'}, 42);
        remock.clearMocks();
        expect(test.key).toBe('key');
        expect(test.props).toEqual({
          prop1: 1,
          prop2: 2,
          children: 42
        })
      });

      it('replace child', () => {
        remock.mock('ComponentRed', () => ({children: 24 as any}));
        const test = React.createElement(ComponentRed, {prop1: 1, children: 22}, 42);
        remock.clearMocks();
        expect(test.props).toEqual({
          prop1: 1,
          children: 24
        })
      });

      it('replace child via props', () => {
        remock.mock('ComponentRed', () => ({props: {children: 24 as any}}));
        const test = React.createElement(ComponentRed, {prop1: 1}, 42);
        remock.clearMocks();
        expect(test.props).toEqual({
          children: 24
        })
      });

      it('support child overwrite', () => {
        const test1 = React.createElement(ComponentRed, {prop1: 1}, 42);
        expect(test1.props).toEqual({
          prop1: 1,
          children: 42
        });
        const test2 = React.createElement(ComponentRed, {prop1: 2, children: 24}, 42);
        expect(test2.props).toEqual({
          prop1: 2,
          children: 42
        });
      });
    });

    describe('cloneElement', () => {

      const base = React.createElement(ComponentRed, {prop1: 1, key: 'key'}, 42);

      it('replace type', () => {
        remock.mock('ComponentRed', () => ({type: ComponentBlue}));
        const test = React.cloneElement(base, {prop2: 2}, 24);
        remock.clearMocks();
        expect(test.type).toBe(ComponentBlue);
        expect(test.key).toBe('key');
        expect(test.props).toEqual({
          prop1: 1,
          prop2: 2,
          children: 24
        });
      });

      it('replace prop', () => {
        remock.mock('ComponentRed', () => ({props: {prop3: 2}}));
        const test = React.cloneElement(base, {prop2: 2}, 24);
        remock.clearMocks();
        expect(test.props).toEqual({
          prop3: 2,
          children: 24
        })
      });

      it('replace prop, but not children', () => {
        remock.mock('ComponentRed', () => ({props: {prop3: 2}}));
        const test = React.cloneElement(base, {prop2: 2});
        remock.clearMocks();
        expect(test.props).toEqual({
          prop3: 2,
          children: 42
        })
      });

      it('merge prop', () => {
        remock.mock('ComponentRed', (type, props) => ({props: {...props, prop3: 3}, children: 24 as any}));
        const test = React.cloneElement(base, {prop2: 2}, 24);
        remock.clearMocks();
        expect(test.props).toEqual({
          prop1: 1,
          prop2: 2,
          prop3: 3,
          children: 24
        })
      });

      it('replace prop', () => {
        remock.mock('ComponentRed', () => ({children: 24 as any}));
        const test = React.cloneElement(base, {prop2: 2}, 24);
        remock.clearMocks();
        expect(test.props).toEqual({
          prop1: 1,
          prop2: 2,
          children: 24
        })
      });
    });

  });

});
