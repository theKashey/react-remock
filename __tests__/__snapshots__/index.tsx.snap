// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remock mock should unwrap renderprop 1`] = `
"<MockedTransparent itWas=\\"RenderProp\\">
  <div>
    24
  </div>
</MockedTransparent>"
`;
